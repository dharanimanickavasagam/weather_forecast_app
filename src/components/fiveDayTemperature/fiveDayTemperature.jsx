import React, { Component } from "react";
import Table from "../common/table";
import { getFiveDayTemperature } from "../../services/fiveDayTemperatureService";
import { kelvinToFahrenheitConverter } from "../../utils/kelvinToFahrenheitConverter";
import { formatDateTime } from "../../utils/formatDateTime";
import Header from "../common/header";
import _ from "lodash";

class FiveDayTemperature extends Component {
  state = {
    cityName: "",
    data: [],
    columns: ["Date", "Time", "Temperature Â°F", "Weather Description", ""]
  };

  async componentDidUpdate(prevProps, prevState) {
    if (
      prevProps.latitude !== this.props.latitude ||
      prevProps.longitude !== this.props.longitude
    ) {
      this.getFiveDayTemperatureAtCurrentLocation();
    }
  }

  getFiveDayTemperatureAtCurrentLocation = async () => {
    try {
      const { city, list } = await getFiveDayTemperature(
        this.props.latitude,
        this.props.longitude
      );
      const { name: cityName } = city;
      let weatherData = {};

      const data = list.map(item => {
        weatherData = {};

        weatherData["date"] = formatDateTime(item.dt_txt, "dddd, MM-DD-YYYY");
        weatherData["time"] = formatDateTime(item.dt_txt, "HH:mm");
        weatherData["temp"] = (
          <b> {kelvinToFahrenheitConverter(item.main.temp)}</b>
        );
        weatherData["description"] = _.upperFirst(item.weather[0].description);
        weatherData["icon"] = (
          <img
            src={`http://openweathermap.org/img/wn/${item.weather[0].icon}@2x.png`}
            alt="weather_icon"
          />
        );
        return weatherData;
      });

      this.setState({ cityName, data });
    } catch (error) {
      alert("Something went wrong..");
    }
  };

  render() {
    return (
      <div className="container">
        <Header name={this.state.cityName} variant={"h6"} />
        <Table data={this.state.data} columns={this.state.columns} />
      </div>
    );
  }
}

export default FiveDayTemperature;
