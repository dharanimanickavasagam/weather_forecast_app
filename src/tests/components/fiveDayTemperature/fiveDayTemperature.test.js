import React from "react";
import { shallow, mount } from "enzyme";
import Header from "../../../components/common/header";
import Table from "../../../components/common/table";
import FiveDayTemperature from "../../../components/fiveDayTemperature/fiveDayTemperature";
import { data } from "../../fixures/fiveDayTemperature";

let wrapper, jsdomAlert;

beforeEach(() => {
  wrapper = shallow(<FiveDayTemperature />);
});

beforeAll(() => {
  jsdomAlert = window.alert;
  window.alert = () => {};
});

describe("it renders correctly", () => {
  test("it should render ", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("it displays the 5day forecast", () => {
  test("it should render Header and Table after the data is loaded", () => {
    expect(wrapper.containsMatchingElement(<Header />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<Table />)).toEqual(true);
  });

  test("it should render the cityname in the header", () => {
    wrapper.setState({ cityName: "Detroit" });
    expect(wrapper.find("Header").prop("name")).toEqual("Detroit");
  });

  test("it should render the table with the forecast data", () => {
    wrapper.setState({ data });
    expect(wrapper.find("Table").prop("data")).toEqual(data);
  });
});

describe("testing the method componentDidUpdate", () => {
  it("trigger a componentDidUpdate when the prevProps and the currentProps are different", async () => {
    const mockReturnValue = {
      city: "Detroit",
      list: [
        {
          dt: 1572188400,
          main: {
            temp: 285.49,
            temp_min: 285.49,
            temp_max: 285.86,
            pressure: 1002,
            sea_level: 1002,
            grnd_level: 972,
            humidity: 86,
            temp_kf: -0.37
          },
          weather: [
            { id: 500, main: "Rain", description: "light rain", icon: "10d" }
          ],
          clouds: { all: 98 },
          wind: { speed: 10.92, deg: 269 },
          rain: { "3h": 0.5 },
          sys: { pod: "d" },
          dt_txt: "2019-10-27 15:00:00"
        }
      ]
    };
    const getFiveDayTemperature = jest
      .fn()
      .mockReturnValue(() => mockReturnValue);

    const prevProps = {
      latitude: null,
      longitude: null
    };

    expect(wrapper).toMatchSnapshot();
    wrapper = shallow(<FiveDayTemperature {...prevProps} />);
    wrapper.setProps({ latitude: 12, longitude: 12 });

    //expect(getFiveDayTemperature).toHaveBeenCalled();

    // const mockGetFiveDayTemperatureAtCurrentLocation = jest.fn();
    // FiveDayTemperature.prototype.getFiveDayTemperatureAtCurrentLocation = mockGetFiveDayTemperatureAtCurrentLocation.mockResolvedValue(
    //   () => wrapper.setState({ cityName: "Detroit", data: mockData })
    // );

    // expect(wrapper).toMatchSnapshot();
    // mockGetFiveDayTemperatureAtCurrentLocation();
    // expect(mockGetFiveDayTemperatureAtCurrentLocation).toHaveBeenCalled();
  });
});
