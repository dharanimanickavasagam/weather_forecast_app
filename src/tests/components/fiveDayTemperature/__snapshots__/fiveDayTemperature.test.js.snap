// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly it should render  1`] = `
<div
  className="container"
>
  <Header
    name=""
    variant="h6"
  />
  <Table
    columns={
      Array [
        "Date",
        "Time",
        "Temperature °F",
        "Weather Description",
        "",
      ]
    }
    data={Array []}
  />
</div>
`;

exports[`testing the method componentDidUpdate trigger a componentDidUpdate when the prevProps and the currentProps are different 1`] = `
<div
  className="container"
>
  <Header
    name=""
    variant="h6"
  />
  <Table
    columns={
      Array [
        "Date",
        "Time",
        "Temperature °F",
        "Weather Description",
        "",
      ]
    }
    data={Array []}
  />
</div>
`;
